File: Heap/leetcode/855_Exam_Room/855_Exam_Room.py
'''
Exam Room
- n개의 좌석이 있고,
- 아무도 없을 때, 1번째부터 채우기
- 사람이 들어오면 가장 가까운 사람으로부터 가장 멀게 앉아야 함. : 0 -> n -> n//2
- 앉을 수 있는 자리가 여러 자리면, 가장 낮은 숫자부터 앉기.


ExamRoom(int n) 
- Initializes the object of the exam room with the number of the seats n.
int seat() 
- Returns the label of the seat at which the next student will set.
void leave(int p) 
- Indicates that the student sitting at seat p will leave the room. 
- It is guaranteed that there will be a student sitting at seat p.

'''

import bisect
##binary ssearch를 활용해서 검색 혹은 삽입
'''
bisect.bisect(a, x, lo=0, hi= len(a)) : a에 x 값이 들어갈 자리의 인덱스 값을 반환
bisect.bisesct_right(a, x) : x와 동일한 값이 존재하면 해당 값의 뒷 인덱스 값을 반환
bisect.bisect_left(a, x) : x와 동일한 값의 인덱스 반환
bisect.insort(a, x) : 정렬된 시퀀스 a에 x값을 삽입
'''


class ExamRoom:

    def __init__(self, n: int):
        self.n = n
        self.students = [] # 현재 앉아 있는 좌석 정보
        

    def seat(self) -> int:
        '''
        가장 가까운 사람과 가장 멀어지기
        '''
        if not self.students :  # 빈 것이라면 맨 처음
            student  = 0
        
        else:
            #초기값 : 그냥 처음엔 혼자 앉아있으니까.
            max_dist = self.students[0]
            student = 0


            
            for i in range(1, len(self.students)): 
                # 앉은 자리 사이사이 공간 넓이를 체크 -> 가장 넓은 공간의 중앙에 앉아버리기

                prev = self.students[i -1]
                curr = self.students[i]

                dist = (curr- prev) // 2
                
                if dist > max_dist: # 가장 넓은지 체크
                    max_dist = dist #여기다!
                    student = prev + dist # 에잇 앉아버려!
    
            #마지막 칸에 앉는 경우도 생각하기
            if (self.n-1) -(self.students[-1])> max_dist :
                student = self.n-1
        
        bisect.insort(self.students, student)
        return student
        '''
        위 코드의 이슈 : 결국 매번 들어올때마다 처음부터 검사해야 함.
        '''
        

    def leave(self, p: int) -> None:
        self.students.remove(p)


# Your ExamRoom object will be instantiated and called as such:
# obj = ExamRoom(n)
# param_1 = obj.seat()
# obj.leave(p)

File: Heap/leetcode/855_Exam_Room/855_Exam_Room.c
#include<stdio.h>
#include<stdlib.h>

/*
동적 할당과
struct 사용 및 array사용에 익숙해질 것!

*/

typedef struct{
    int n;
    int* seats; // 정렬된 좌석
    int seat_count;
} ExamRoom;



// 배열 삽입 및, 정렬 상태 유지
void insert_seat(ExamRoom* room, int seat){
    int i = room->seat_count-1; // 끝에서부터


    while(i>=0 && room->seats[i]>seat){
        room->seats[i+1] = room->seats[i]; // 한칸씩 뒤로 미루기.
        i--;
    }

    //위치 찾았다면 넣기 (i는 seat보다 작은 것을 찾은 것의 인덱스)
    room->seats[i+1] = seat;
    room->seat_count++;
}

void remove_seat(ExamRoom* room, int seat){
    // 특정 값 제거

    int i = 0;
    // 찾아서 제거
    while(i<room->seat_count && room->seats[i] != seat){
        i++;
    }
    //아예 끝 혹은 찾은 인덱스
    if(i<room->seat_count){
        //찾은 경우
        //제거 하고 이 이후꺼 다 앞으로 가져와야 함.
        for(int j = i; j<room->seat_count;j++){
            room->seats[j] = room->seats[j+1]; //뒤에거 앞으로 당기기
        }
        room->seat_count --;
    }

}

// 객체 생성
ExamRoom* examRoomCreate(int n) {
    ExamRoom* room = (ExamRoom*)malloc(sizeof(ExamRoom)); // 동적 할당
    room-> n= n;
    room->seats = (int*)malloc(n*sizeof(int)); // 최대 n개의 배열
    room ->seat_count = 0;
    return room;
}


//앉히기
int examRoomSeat(ExamRoom* obj) {
    
    int seat; // 앉을 위치
    if(obj->seat_count ==0){
        seat = 0;
    }

    else{
        int max_dist = obj->seats[0];
        seat = 0;


        for(int i =1; i<obj->seat_count ; ++i){
            int prev = obj->seats[i-1];
            int curr = obj->seats[i];

            int dist = (curr - prev)/2;

            if(dist>max_dist){
                max_dist = dist;
                seat = prev + dist;
            }
        }

        if((obj->seat_count -1 - obj->seats[-1]) > max_dist)
        {
            seat = obj->seat_count -1;
        }
    }
    insert_seat(obj, seat);

    return seat;
}

//떠나기
void examRoomLeave(ExamRoom* obj, int p) {
    remove_seat(obj, p);
}


//객체 제거
void examRoomFree(ExamRoom* obj) {
    free(obj->seats);
    free(obj);
}


